source/RemoteWiring/HardwareProfile.h

```diff
@@ -0,0 +1,351 @@
+    ///</summary>
+    bool
+    isAnalogSupported(
+        uint8_t pin_
+        );
+
+    ///<summary>
+    ///returns true if the digital input capability is supported by the given pin number
+    ///<param name="pin_">The requested pin</param>
+    ///<returns>true if the pin and this hardware profile are both valid and the pin supports the digital input capability, false otherwise</returns>
+    ///</summary>
+    bool
+    isDigitalInputSupported(
+        uint8_t pin_
+        );
+
+    ///<summary>
+    ///returns true if the digital input pullup capability is supported by the given pin number
+    ///<param name="pin_">The requested pin</param>
+    ///<returns>true if the pin and this hardware profile are both valid and the pin supports the digital input pullup capability, false otherwise</returns>
+    ///</summary>
+    bool
+    isDigitalInputPullupSupported(
+        uint8_t pin_
+        );
+
+    ///<summary>
+    ///returns true if the digital output capability is supported by the given pin number
+    ///<param name="pin_">The requested pin</param>
+    ///<returns>true if the pin and this hardware profile are both valid and the pin supports the digital output capability, false otherwise</returns>
+    ///</summary>
+    bool
+    isDigitalOutputSupported(
+        uint8_t pin_
+        );
+
+    ///<summary>
+    ///returns true if the I2C capability is supported by the given pin number
+    ///<param name="pin_">The requested pin</param>
+    ///<returns>true if the pin and this hardware profile are both valid and the pin supports the I2C capability, false otherwise</returns>
+    ///</summary>
+    bool
+    isI2cSupported(
+        uint8_t pin_
+        );
+
+    ///<summary>
+    ///returns true if the PWM capability is supported by the given pin number
+    ///<param name="pin_">The requested pin</param>
+    ///<returns>true if the pin and this hardware profile are both valid and the pin supports the PWM capability, false otherwise</returns>
+    ///</summary>
+    bool
+    isPwmSupported(
+        uint8_t pin_
+        );
+
+    ///<summary>
+    ///returns true if the servo capability is supported by the given pin number
+    ///<param name="pin_">The requested pin</param>
+    ///<returns>true if the pin and this hardware profile are both valid and the pin supports the servo capability, false otherwise</returns>
+    ///</summary>
+    bool
+    isServoSupported(
+        uint8_t pin_
+        );
+
+private:
+    std::atomic_bool _is_valid;
+
+    //stateful members received from the device
+    std::atomic_int _analog_offset;
+    std::atomic_int _num_analog_pins;
+    std::atomic_int _total_pins;
+    std::vector<uint8_t> *_pinCapabilities;
+    //for each of the following maps: K = pin number, V = resolution value in bits
+    std::map<uint8_t, uint8_t> *_analogResolutions;
```

Comment:
std::array can't be initialized without declaring the size ahead of time, as a parameter to the generic type. We don't know the type, so a vector or map seemed most appropriate. I like map as it is guaranteed O(1) lookup time. Once the map is created, we should never again perform add or remove, only many many lookups